pandas/core/tools/datetimes.py

```diff
@@ -514,18 +519,41 @@ def _convert_listlike(arg, box, format, name=None, tz=tz):
             arg = np.asarray(arg)
         arg = arg + offset
 
+    convert_cache = None
+    if cache and is_list_like(arg) and len(arg) >= 1000:
+        unique_dates = algorithms.unique(arg)
+        if len(unique_dates) != len(arg):
+            from pandas import Series
+            cache_dates = _convert_listlike(unique_dates, False, format)
+            convert_cache = Series(cache_dates, index=unique_dates)
```

Comment:
so its better to actually to make ``convert_cache`` a function, which can then take the unique_dates and return the converted data, avoids lots of code duplication.